docs/design/values.md

```diff
@@ -0,0 +1,920 @@
+declarations implicitly starting off within a `var` pattern while `let`
+declarations introduce patterns that work the same as function parameters and
+others with bindings that produce values by default.
+
+### Consuming function parameters
+
+Just as part of a `let` binding can use a `var` prefix to become a variable
+pattern and bind names that will form reference expressions to the variable's
+storage, so can function parameters:
+
+```carbon
+fn Consume(var x: SomeData) {
+  // We can mutate and use variable that `x` refers to here.
+}
+```
+
+This allows us to model an important special case of function inputs -- those
+that are _consumed_ by the function, either through local processing or being
+moved into some persistent storage. Marking these in the pattern and thus
+signature of the function changes the expression category of arguments in the
+caller, causing them to become _initializing expressions_ that directly
+initialize storage dedicated-to and owned-by the function parameter.
+
+This pattern serves the same purpose as C++'s pass-by-value when used with types
+that have non-trivial resources attached to pass ownership into the function and
+consume the resource. But rather than that being the seeming _default_, Carbon
+makes this a use case that requires a special marking on the declaration.
+
+## Reference expressions
+
+_Reference expressions_ refer to _objects_ with _storage_ where a value may be
+read or written and the object's address can be taken. There are two
+sub-categories of reference expressions: _durable_ and _ephemeral_.
+
+Calling a [method](/docs/design/classes.md#methods) on a reference expression
+where the method's `self` parameter has an `addr` specifier can always
+implicitly take the address of the referred-to object. This address is passed as
+a [pointer](#pointers) to the `self` parameter for such methods.
+
+### Durable reference expressions
+
+_Durable reference expressions_ are those where the object's storage outlives
+the full expression and the address could be meaningfully propagated out of it
+as well.
+
+There are two expressions that require one of their operands to be a durable
+reference expression in Carbon:
+
+-   [Assignment expressions](/docs/design/assignment.md) require the
+    left-hand-side of the `=` to be a durable reference. This stronger
+    requirement is enforced before the expression is rewritten to dispatch into
+    the `Carbon.Assign.Op` interface method.
+-   [Address-of expressions](#pointer-syntax) require their operand to be a
+    durable reference and compute the address of the referenced object.
+
+There are several kinds of expressions that produce durable references in
+Carbon:
+
+-   Names of objects introduced with a
+    [variable binding](#binding-patterns-and-local-variables-with-let-and-var):
+    `x`
+-   Dereferenced [pointers](#pointers): `*p`
+-   Names of subobjects through member access: `x.member` or `p->member`
```

Comment:
This only applies if `x` is a durable reference expression, right?

Reply:
Yes, clarified I think.